SEP-2549: TTL for List Results (#2549)
modelcontextprotocol/modelcontextprotocol: seps/2549-TTL-for-list-results.md

# SEP-2549: TTL for List Results

- **Status**: Draft
- **Type**: Standards Track
- **Created**: 2026-04-09
- **Author(s)**: Caitie McCaffrey (@CaitieM20)
- **Sponsor**: @CaitieM20 
- **PR**: https://github.com/modelcontextprotocol/specification/pull/2549

## Abstract

> This SEP proposes adding an optional `ttl` (time-to-live) field to the result objects returned by `tools/list`, `prompts/list`, `resources/list`, and `resources/templates/list`. The TTL tells clients how long the response may be considered fresh before re-fetching. This allows clients to cache feature lists and poll on a predictable schedule, reducing reliance on server-push `list_changed` notifications while remaining fully backward compatible. TTL supplements rather than replaces the existing notification mechanism — both can coexist.

Reviewer: The Abstract says TTL lets clients "poll on a predictable schedule," but Freshness calculation below says clients **SHOULD NOT** treat TTL as a polling interval.

```suggestion
This SEP proposes adding an optional `ttl` (time-to-live) field to the result objects returned by `tools/list`, `prompts/list`, `resources/list`, and `resources/templates/list`. The TTL tells clients how long the response may be considered fresh before re-fetching. This allows clients to cache feature lists and lazily revalidate them on access, reducing reliance on server-push `list_changed` notifications while remaining fully backward compatible. TTL supplements rather than replaces the existing notification mechanism, and both can coexist.
```

Author: good call out updated the wording here to just remove the polling bit. The real thing is caching & eliminating the need for notifications

4. **Alignment with web standards**: HTTP caching (`Cache-Control: max-age`) and DNS TTLs have long demonstrated that time-based freshness hints are a simple, well-understood mechanism for reducing unnecessary refetches. MCP can benefit from the same pattern.

Adding a TTL field to list responses solves all of these problems with a minimal, backward-compatible protocol change.

## Specification

### New field: `ttl`

An optional `ttl` field is added to the base `PaginatedResult` interface. Because `ListToolsResult`, `ListPromptsResult`, `ListResourcesResult`, and `ListResourceTemplatesResult` all extend `PaginatedResult`, the field is inherited by all four list result types.

#### Schema change (TypeScript)

```typescript
/** @internal */
export interface PaginatedResult extends Result {
  /**
   * An opaque token representing the pagination position after the last
   * returned result. If present, there may be more results available.
   */
  nextCursor?: Cursor;

  /**
   * An optional hint from the server indicating how long (in seconds) the
   * client MAY cache this response before re-fetching. Semantics are

>    * analogous to HTTP Cache-Control max-age.

Reviewer: Minor: the `ttl=0` semantics defined below ("SHOULD NOT serve a cached copy") match DNS TTL=0 rather than HTTP `max-age=0`, which per [RFC 9111 Section 4.2.1](https://www.rfc-editor.org/rfc/rfc9111#section-4.2.1) means immediately-stale-but-storable (and thus eligible for `stale-if-error`). "Loosely analogous to HTTP `max-age` and DNS TTL" would be more precise, but this is a nit.

Author: we want this to be treated like max-age=0, Will update the comments to make this clearer

   *
   * - If absent, the client has no server-provided freshness guidance and
   *   SHOULD rely on notifications or its own heuristics.
   * - If 0, the client SHOULD re-fetch every time the list is needed and
   *   SHOULD NOT serve a cached copy.
   * - If positive, the client SHOULD consider the list fresh for this many
   *   seconds after receiving the response. The client SHOULD NOT re-fetch
   *   before the TTL expires unless it receives a list_changed notification.
   */
  ttl?: number;
}
```

> **Open Question — TTL format**: An alternative representation is an ISO 8601 duration string (e.g., `"PT5M"` for 5 minutes). Integer seconds are simpler, consistent with HTTP `max-age`, and easier to compare arithmetically. ISO 8601 durations are more human-readable and used in some Azure/AWS APIs. Community input is welcome on which format to adopt. The remainder of this specification uses integer seconds for illustration.

> ### Semantics

Reviewer: All the SHOULDs in this section land on the client. What's the server actually committing to when it sends `ttl: 300`? If it doesn't advertise `listChanged` and then changes the list at t=100, is that a spec violation or expected? The word "hint" and the `max-age` analogy suggest it's fine, but neither is normative, and meanwhile we're telling clients they SHOULD NOT re-fetch early.

> A TTL is a freshness estimate, not a guarantee. Servers MAY change the underlying list before the TTL expires; servers that do so and have advertised `listChanged` SHOULD send the corresponding notification. Clients accept that cached data may be up to `ttl` seconds stale.

Also gives server authors a starting point for "how do I pick a value."

Author: Great suggestion, I've added this. 

And correct we want this to be a hint not a binding contract. Also I've removed the language about not re-fetching before the ttl, since clients may want to if they get an error from the tool call indicating that their list is stale for instance

| Condition                                                    | Client behavior                                                                                      |
| ------------------------------------------------------------ | ---------------------------------------------------------------------------------------------------- |
| `ttl` absent                                                 | No freshness hint. Client SHOULD rely on `list_changed` notifications or poll at its own discretion. |
| `ttl` = 0                                                    | Do not cache. Client SHOULD re-fetch every time the list is needed.                                  |
| `ttl` > 0                                                    | Client SHOULD consider the response fresh for `ttl` seconds from receipt. Client SHOULD NOT re-fetch before the TTL expires. |
| `list_changed` notification received while TTL is active     | The notification invalidates the cached response. Client SHOULD re-fetch regardless of remaining TTL. |

#### Freshness calculation

> A client records the local time at which the response was received (`t_received`). The response is considered **fresh** while `now < t_received + ttl`. Once the TTL expires the response is **stale** and the client SHOULD re-fetch on next access.

Reviewer: What happens after expiry if the re-fetch fails (server unreachable, transport error)? HTTP needed `must-revalidate` vs `stale-if-error` to disambiguate exactly this, and without guidance one SDK will fall back to stale data while another surfaces an empty list. This is the inverse of the pre-expiry case @markdroth raised above.

```suggestion
A client records the local time at which the response was received (`t_received`). The response is considered **fresh** while `now < t_received + ttl`. Once the TTL expires the response is **stale** and the client SHOULD re-fetch on next access. If the re-fetch fails or the server is unreachable, the client MAY continue to serve the stale cached response and SHOULD surface staleness to the host application.
```

Reviewer 2: I found this PR really interesting, working on an MCP server generator; it feels like this direction could be really helpful for users configuring across multiple hosts.

I had a quick question on the host policy section. The spec says clients SHOULD respect the server's TTL.
When a server provides something like `ttl: 3600`, is it expected that a host can enforce a stricter effective TTL?
But what if:
- Cursor wants to cap caching at 5 minutes for freshness?
- Claude wants no caching at all for sensitive tools?
- A wrapper re-fetching on auth/session changes
- A gateway wants to enforce max TTL = 1 hour?

Should we document that **hosts MAY apply a stricter effective TTL** 
(e.g., `min(server.ttl, host.maxTTL)`)?

I'm happy to document this section or explore how SDKs currently handle TTL implementation if that’s useful.
cc: @localden @CaitieM20

Author: Updated the wording in the SEP to say "Clients MAY serve stale responses if errors occur in re-fetching results(e.g., network issues, server downtime). The TTL is a hint for how long the client can safely rely on the data, but real-world conditions may require flexibility."

This seems like the right call for graceful degradation, for example the tool calls that haven't changed will still work in the interim.

Author: @madhaviai I've updated some of the spec language to focus on the assumptions around freshness and less on the expected client behavior, because clients/hosts may have different needs/guarantees they want to provide as you call out above.

We also have this line "Clients MAY re-fetch if they have reason to believe the data has changed, even if the TTL has not yet expired. Examples include receiving an unexpected error on a tool call indicating that the the method was not found or the parameters were invalid."

Clients can essentially purge their caches whenever they want or have unexpected failures where they crash and therefore need to reboot and repopulate their caches and therefore can always call the methods as needed.  That's why all this language is MAY and SHOULD.

Reviewer 2: @CaitieM20 Makes sense, TTL as a flexible hint 
with MAY/SHOULD language makes sense for real-world scenarios.

Clients SHOULD NOT treat TTL as a polling interval that triggers automatic background refetches. The TTL is a **freshness hint**: the client checks freshness when it needs the list, and re-fetches only if stale. Implementations that do choose to poll SHOULD apply jitter and backoff.

### Interaction with `list_changed` notifications

TTL and `list_changed` notifications are complementary:

- A server MAY provide `ttl` without advertising `listChanged: true` in its capabilities. In this case the client relies entirely on TTL. 
- A server MAY advertise `listChanged: true` **and** provide `ttl`. In this case the client can use the TTL to avoid unnecessary refetches between notifications, and the notification acts as an immediate invalidation signal.
- A server MAY advertise `listChanged: true` without providing `ttl`. Behavior is unchanged from today.

> - A server SHOULD use one of the mechanisms (TTL or notifications) rather than neither, to ensure clients have some way to stay up to date.

Reviewer: This new SHOULD makes a server that today offers neither TTL nor `listChanged` SHOULD-non-conformant, which sits a bit awkwardly with "No existing fields or behaviors are modified" in Backward Compatibility below. Since this bullet isn't load-bearing for the design:

```suggestion
- Servers are encouraged to use at least one of the mechanisms (TTL or notifications) so clients have some way to stay up to date.
```

```mermaid
sequenceDiagram
    participant C as Client
    participant S as Server

    C->>S: tools/list
    S-->>C: { tools: [...], ttl: 300 }
    Note over C: Cache response, fresh for 5 min

    Note over C: 2 minutes later...
    C->>C: Need tools list → cache still fresh, use cached

    Note over C: 3 minutes later (TTL expired)...
    C->>C: Need tools list → cache stale
    C->>S: tools/list
    S-->>C: { tools: [...], ttl: 300 }

    Note over S: Tools change before TTL expires
    S-->>C: notifications/tools/list_changed
    Note over C: Invalidate cache immediately
    C->>S: tools/list
    S-->>C: { tools: [...], ttl: 300 }
```

### Interaction with pagination

When a list result includes `nextCursor` (indicating more pages), the `ttl` applies to the **entire paginated list**, not to individual pages. Specifically:

> - The TTL SHOULD only appear on every page with the same value. Clients SHOULD use the TTL from the last page they fetched to determine freshness.

Reviewer: "SHOULD only appear on every page" reads two ways.

```suggestion
- Servers that include `ttl` SHOULD include it on every page of a paginated response and SHOULD use the same value on each page. Clients SHOULD use the TTL from the last page they fetched to determine freshness.
```

Author: updated this section

- When the TTL expires, the client SHOULD re-fetch from the beginning (without a cursor) to get the full updated list.

### No new capability flag

No new capability flag is needed. The `ttl` field is optional on the response object. Servers that do not wish to provide a TTL simply omit the field. Clients that do not understand the field ignore it per standard JSON handling of unknown properties.

### Error handling

- If `ttl` is present but is not a non-negative integer, the client SHOULD ignore it and behave as if it were absent.
- Clients MUST NOT treat a missing `ttl` as an implicit TTL of 0 or any other value.

## Rationale

### Why add TTL to `PaginatedResult` rather than each result type individually?

All four list result types extend `PaginatedResult`. Adding the field there avoids repetition and ensures any future list result types also inherit TTL support. This is consistent with how `nextCursor` is already defined.

> `ListTasksResult` also inherits from `PaginatedResult`, but in the move to stateless transports this feature will be removed. So for simplicity we are adding it to `PaginatedResult` now and will remove it from `ListTaskResult` in a future SEP.

Reviewer: ```suggestion
`ListTasksResult` also inherits from `PaginatedResult`, but in the move to stateless transports this feature will be removed. So for simplicity we are adding it to `PaginatedResult` now and will remove it from `ListTasksResult` in a future SEP.
```

| GraphQL `@cacheControl`        | `maxAge` integer secs  | Per-field cache hints in GraphQL responses                         |
| gRPC `grpc-retry-pushback-ms`  | Milliseconds           | Server-provided retry hint (different use case, similar pattern)   |

Integer seconds is the most common representation across these systems.

### Why not use HTTP caching directly?

MCP is transport-agnostic. While HTTP-based transports could theoretically use `Cache-Control` headers, MCP also operates over stdio,  and supports pluggable transports where HTTP headers may not be available. Embedding the TTL in the JSON response body ensures it works uniformly across all transports.

## Backward Compatibility

This change is fully backward compatible:

- The `ttl` field is optional. Existing servers that do not provide it continue to work unchanged.
- Existing clients that do not understand the field will ignore it, as MCP result objects permit additional properties via `[key: string]: unknown` on the `Result` base type.
- No existing fields or behaviors are modified or removed.
- No capability negotiation is required.

> ## Reference Implementation

Reviewer: Missing a Security Implications section before this (required by [SEP guidelines](https://modelcontextprotocol.io/community/sep-guidelines) and the [SEP template](https://github.com/modelcontextprotocol/modelcontextprotocol/blob/main/seps/TEMPLATE.md)). The one MCP-specific point worth noting is that a misconfigured or malicious server could set a very large TTL to suppress client refresh, so clients MAY cap the honored TTL at an implementation-defined maximum.
